library/std/src/sys/exit_guard.rs

```diff
@@ -31,27 +25,29 @@ cfg_if::cfg_if! {
         #[cfg_attr(any(test, doctest), allow(dead_code))]
         pub(crate) fn unique_thread_exit() {
             let this_thread_id = unsafe { libc::pthread_self() };
-            use crate::sync::{Mutex, PoisonError};
-            static EXITING_THREAD_ID: Mutex<Option<PThread>> = Mutex::new(None);
-            let mut exiting_thread_id =
-                EXITING_THREAD_ID.lock().unwrap_or_else(PoisonError::into_inner);
-            match *exiting_thread_id {
-                None => {
+
+            const _: () = assert!(mem::size_of::<libc::pthread_t>() <= mem::size_of::<usize>());
+
+            const NONE: usize = 0;
+            static EXITING_THREAD_ID: AtomicUsize = AtomicUsize::new(NONE);
+            match EXITING_THREAD_ID.compare_exchange(NONE, this_thread_id as usize, Ordering::Relaxed, Ordering::Relaxed).unwrap_or_else(|id| id) {
+                NONE => {
                     // This is the first thread to call `unique_thread_exit`,
                     // and this is the first time it is called.
-                    // Set EXITING_THREAD_ID to this thread's ID and return.
-                    *exiting_thread_id = Some(PThread(this_thread_id));
-                },
-                Some(exiting_thread_id) if exiting_thread_id.0 == this_thread_id => {
+                    //
+                    // We set `EXITING_THREAD_ID` to this thread's ID already
+                    // and will return.
+                }
+                id if id == this_thread_id as usize => {
```

Comment:
That's a fuzzy-provenance cast, at least on musl. Please use `AtomicPtr` instead.

Reply:
Yes. Casting a pointer to an integer is a [side-effecting operation](https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html). Also, miri will warn against this once it supports `atexit`, since the exposed provenance model isn't finalized yet.

Reply:
Is casting an arbitrary integer to an (invalid) pointer value a problem under this model? `pthread_t` is an integer on some platforms and a pointer on others.

Reply:
You could use the strict provenance APIs. I.e. [`addr`](https://doc.rust-lang.org/std/primitive.pointer.html#method.addr-1)